RN35/yacp
# LAB-378: Hook up Database; Set up Auth; UI Changes

# Proposed Changes

- Removed old files from `create-react-app` - `client` and `backend` folders
  - This should have been done in the previous PR
- Hooked up a Postgres database
- Using a Prisma as a ORM (https://www.prisma.io/nextjs)
- Set up auth using NextAuth.js (https://next-auth.js.org/adapters/prisma)
   - Currently only supporting auth using GitHub (https://next-auth.js.org/providers/github)
- UI changes (see screenshots) 

# Jira Stories

- [LAB-378](https://oregonstate-innovationlab.atlassian.net/browse/LAB-378)

# Checklist

- [x] This has been tested
- [x] Self-documenting code
- [x] Code is linted

# Instructions to review
## How to Run
- Clone the repository: `git clone git@github.com:RN35/yacp.git`
- Change dir: `cd yacp`
- Checkout branch feat/lab-378: `git checkout feat/lab-378`
- Create a `.env` file in the root and either:
   - Setup db locally and create your own GitHub app :
       - spin up a database locally on docker: `docker run --rm --name vivadb -v viva-volume:/var/lib/postgresql/data -e POSTGRES_PASSWORD=viva -p 5432:5432 -d postgres:latest`
       - to create `GITHUB_ID` and `GITHUB_SECRET` refer https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/authorizing-oauth-apps
       - remember that if you choose this approach you will have to make entries in the database yourself. To do that run `npx prisma studio`. (https://www.prisma.io/studio) - Already installed as a dev dependency. 
       - 
![image](https://user-images.githubusercontent.com/15828512/224526350-92a75fc9-783b-435e-8275-536cc5d9166c.png)

   - Or use the DEV setup I am using (don't worry the repo is private and I will change these in the future):
       - Postgres is deployed on Supabase (https://supabase.com/)
```
# .env
DATABASE_URL="postgres://postgres:GXM7QvCxzFtJN4XX@db.rgivfocetfnvucmhgnxv.supabase.co:5432/postgres"


# GitHub OAuth
GITHUB_ID=da8bb4588f7e05abcb6a
GITHUB_SECRET=4197cf5ca124f2e6dab58d3a62aac6c64f7fbbb9
NEXTAUTH_URL=http://localhost:3000/api/auth
```

- Install dependencies: `npm install`
- Spin up development server: `npm run dev`

The code changes have been deployed to https://yacp.vercel.app/ as well.

## How to Validate
Open `http://localhost:3000` in any web browser. You need to "Sign In" to see the homepage. 

![image](https://user-images.githubusercontent.com/15828512/224526183-e0b37598-1e8c-454e-8201-9e9be000afff.png)

Try going to a valid URL like 'http://localhost:3000/problems/two-sum'. It should throw an error:

![image](https://user-images.githubusercontent.com/15828512/224526213-eea7eed7-3731-47f5-a27a-0c17b7821b1f.png)

Ensure signing in works. 
![image](https://user-images.githubusercontent.com/15828512/224526243-6b9481e5-a182-483d-9727-8c61c7de13e6.png)

And you should be able to see the homepage. 
- Your name and profile picture will be retrieved from GitHub 
- The links are currently hardcoded (so make sure the entries are present in the database) 

![image](https://user-images.githubusercontent.com/15828512/224526259-4b6beead-370a-43a2-921e-83d109d67464.png)

IDE page:
- http://localhost:3000/problems/two-sum
- http://localhost:3000/problems/three-sum

- This page has some UI changes to make it look better
Before:
![image](https://user-images.githubusercontent.com/15828512/224526581-71b55361-3d36-4a7b-895c-fb6ea93c993e.png)

After:
![image](https://user-images.githubusercontent.com/15828512/224526588-9a40a44f-e863-4dfe-ac4b-44feabe6493d.png)


- The description content is saved as `markdown` in the database. This is converted to MUI React components at runtime using https://www.npmjs.com/package/mui-markdown. Check if everything renders properly and looks consistent.
